main/src/library/Iterator.flix

```diff
@@ -500,16 +437,19 @@ namespace Iterator {
     ///
     /// Consumes the entire iterator.
     ///
-    pub def foldRightWithCont(f: (a, Unit -> b \ { ef, Read(r) }) -> b \ ef, z: b, iter: Iterator[a, r]): b \ { ef, Read(r) } =
-        match iter {
-            case Iterator(done, _) if done() => z
-            case Iterator(_, next)           => f(next(), _ -> foldRightWithCont(f, z, iter))
-        }
+    pub def foldRightWithCont(f: (a, Unit -> b \ { ef, Read(r) }) -> b \ ef, z: b, iter: Iterator[a, ef, r]): b \ { ef, Read(r) } =
+        let Iterator(_, iterF) = iter;
+        def loop() = match iterF() {
+            case Skip   => loop()
+            case Ans(a) => f(a, _ -> loop())
+            case Done   => z
+        };
+        loop()
```

Comment:
Its pretty nice that `Iterator` can support `foldRightWithCont` with the right complexity. 

You would expect a corresponds between iterators and foldRightWithCont and here we see that in just a few lines it can be realized.

Reply:
I think that was all Jakob's work and I just removed the `done` thunk.  I still haven't really got an intuition about `foldRightWithCont` as I've not used it much.